### PR TITLE
add skylint check for missing rule documentation

### LINKED ISSUE #4854: skylint does not check missing rules docstring
### Description of the problem / feature request:

skylint does not check for missing docstring of rules.

### Feature requests: what underlying problem are you trying to solve with this feature?

public variables of type rule and repository_rule should throw an error like:
`rule 'foo' has no docstring [missing-rule-docstring]`

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

skylint should not pass on:
```python
"""foo"""

def _impl(ctx):
    print("foo", ctx.attr.bar)

foo = rule(
    _impl,
    attrs = {
        "bar": attr.string(),
    },
)
```

### skylint from Bazel version

0.11.1

### PR DESCRIPTION
**Pull request by ahippler**
close #4854

### CODE REVIEW

**Comment by laurentlb** [comment]
Note that there's now a doc attribute:
```
new_rule = rule(
  implementation = _new_rule_impl,
  attrs = {
    "foo": attr.label(),
    "bar": attr.string(),
  },
  doc = """some documentation"""
)
```

So maybe we should recommend that instead of a docstring?

**Comment by ahippler** [comment]
you are right. I forgot this attribute. I will change it to use the doc attributes of the rules and their attributes.

**Comment by irengrig** [comment]
Hi, Bazel sheriff is here, what is the status of this PR? Should we merge or close it?

File: site/docs/skylark/skylint.md [hunk 1]
```diff
@@ -158,6 +179,7 @@ The analyzer **requires docstrings** for:
 *   each **module** (.bzl file)
 *   each **public function** (i.e. a function not starting with an underscore)
     that contains at least 5 statements
+*   each **public rule definition**
```
Comment (brandjon) [inline site/docs/skylark/skylint.md]: And public attribute of a public rule definition?

**Comment by brandjon** [comment]
Erm, tried to fix the conflict in the contributor file, but it looks like it created a much messier merge commit.

File: site/docs/skylark/skylint.md [hunk 2]
```diff
@@ -139,6 +139,27 @@ def example_function2(foo, bar):
...
+    ),
+  },
+  doc = """Illustrates the format for rule definitions
+
+After the one-line summary comes the description body.
+It contains additional information about the rule.
+""")
 ```
 
 #### Indentation
```
Comment (brandjon) [inline site/docs/skylark/skylint.md]: There should also be a new subsection of the docstrings section describing how to document rules and rule attrs. (Unlike for functions, there's no precise format for the string content, but we need to at least say that the `doc` param should be present.) This new subsection can be the anchor target of the new issue types `missing-rule-documentation` and `missing-attribute-documentation` (see use of `<a>` tags for the existing sections).

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java [hunk 3]
```diff
@@ -274,6 +274,12 @@ private boolean nextLine() {
         line = "";
         return false;
       }
+      if (startOfLineOffset == 0 && docstring.charAt(startOfLineOffset) == '\n') {
+        line = "";
+        endOfLineOffset += 1;
+        return true;
+      }
+
```
Comment (brandjon) [inline src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java]: Why the new newline parse logic?

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java [hunk 4]
```diff
@@ -361,9 +367,15 @@ DocstringInfo parse() {
...
       if (!nextLine()) {
+        if (summary.trim().isEmpty()) {
+          errors.add(new DocstringParseError("the docstring is empty", 1, ""));
+        }
         return new DocstringInfo(
             summary, Collections.emptyList(), "", nonStandardDeprecation, "", null);
       }
+      if (summary.trim().isEmpty()) {
+        errors.add(new DocstringParseError("the docstring is empty", 1, ""));
+      }
```
Comment (brandjon) [inline src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java]: So we're requiring function docstrings to be non-empty, but not rule/attr docstrings?

File: src/tools/skylark/javatests/com/google/devtools/skylark/skylint/DocstringUtilsTest.java [hunk 5]
```diff
@@ -28,6 +28,45 @@
...
+    DocstringInfo info = DocstringUtils.parseDocstring("\n\n", 0, errors);
+    Truth.assertThat(info.summary).isEmpty();
+    Truth.assertThat(info.parameters).isEmpty();
+    Truth.assertThat(info.longDescription).isEmpty();
+    Truth.assertThat(errors.toString())
+      .contains("1: the one-line summary is missing");
+
+
+    errors = new ArrayList<>();
+    info = DocstringUtils.parseDocstring("\nfoo\n", 0, errors);
```
Comment (brandjon) [inline src/tools/skylark/javatests/com/google/devtools/skylark/skylint/DocstringUtilsTest.java]: So this now raises an Issue?

    def foo():
        """
        Does a foo.
        """

This seems like a separate stylistic discussion that should be in another PR. Is there any corresponding github issue / consensus?

Comment (ahippler) [reply src/tools/skylark/javatests/com/google/devtools/skylark/skylint/DocstringUtilsTest.java]: i uses the format of skydoc (https://skydoc.bazel.build/docs/writing.html). the doc format should be matching.

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java (hunk 4 above)
Comment (ahippler) [reply src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java]: i could add a check for empty string  rule/attr docstrings. what di you think?

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringChecker.java [hunk 6]
```diff
@@ -257,7 +348,7 @@ private Issue docstringParseErrorToIssue(StringLiteral docstring, DocstringParse
       startColumn = 1;
     }
     Location start = new Location(startLine, startColumn);
-    Location end = new Location(startLine, Math.max(1, startColumn + error.line.length() - 1));
+    Location end = new Location(startLine, Math.max(startColumn, startColumn + error.line.length() - 1));
```
[the remark this answers is not in the document]
Comment (ahippler) [reply src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringChecker.java]: i think its for the missing one-line summary

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java (hunk 3 above)
Comment (ahippler) [reply src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java]: also for missing one-line summary

File: src/tools/skylark/javatests/com/google/devtools/skylark/skylint/DocstringUtilsTest.java (hunk 5 above)
Comment (ahippler) [reply src/tools/skylark/javatests/com/google/devtools/skylark/skylint/DocstringUtilsTest.java]: @brandjon

**Comment by ulfjack** [comment]
Is this in a state where it can be reviewed / merged? Who's waiting for what?

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java (hunk 3 above)
Comment (brandjon) [reply src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java]: Ok, I see. Can you add a short comment right above the if, explaining that this catches the case where the first line is empty? That'll make it easier to follow this function's logic alongside the above and below comments.

File: src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java (hunk 4 above)
Comment (brandjon) [reply src/tools/skylark/java/com/google/devtools/skylark/skylint/DocstringUtils.java]: If we're going to prohibit empty docstrings then let's be consistent for other kinds of objects.

Can you use the error() helper instead of constructing DocstringParseError directly?

**Comment by laurentlb** [comment]
This PR is obsolete, Skylint has been deleted.

This check should be added to Buildifier instead (https://github.com/bazelbuild/buildtools/).
@vladmos
